aws/resource_aws_athena_database.go

```diff
@@ -38,18 +38,73 @@ func resourceAwsAthenaDatabase() *schema.Resource {
 				Optional: true,
 				Default:  false,
 			},
+			"encryption_key": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"type": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
 		},
 	}
 }
 
+func getResultConfig(d *schema.ResourceData) (*athena.ResultConfiguration, error) {
+	resultConfig := athena.ResultConfiguration{
+		OutputLocation: aws.String("s3://" + d.Get("bucket").(string)),
+	}
+
+	e := d.Get("encryption_key").([]interface{})
+	if len(e) <= 0 {
+		return &resultConfig, nil
+	}
+
+	data := e[0].(map[string]interface{})
+	keyType := data["type"].(string)
+	keyID := data["id"].(string)
+
+	if len(keyType) <= 0 {
+		return nil, fmt.Errorf("An encryption key type is required")
+	}
+
+	if strings.HasSuffix(keyType, "_KMS") && len(keyID) <= 0 {
+		return nil, fmt.Errorf("Key type %s requires a valid KMS key ID", keyType)
```

Comment:
Does the API not provide a helpful error message already? Since this error handling occurs during apply time anyways, we should avoid adding logic the API already handles unless necessary.

Reply:
It does; just felt dirty making an AWS call for known errors. If that’s the pattern, I’ll update though.